packages/quick_actions/quick_actions_android/android/src/main/java/io/flutter/plugins/quickactions/QuickActions.java

```diff
@@ -0,0 +1,184 @@
+import android.graphics.drawable.Icon;
+import android.os.Build;
+import androidx.annotation.ChecksSdkIntAtLeast;
+import androidx.annotation.NonNull;
+import io.flutter.plugins.quickactions.Messages.AndroidQuickActionsApi;
+import io.flutter.plugins.quickactions.Messages.FlutterError;
+import io.flutter.plugins.quickactions.Messages.Result;
+import io.flutter.plugins.quickactions.Messages.ShortcutItemMessage;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+final class QuickActions implements AndroidQuickActionsApi {
+  protected static final String EXTRA_ACTION = "some unique action key";
+
+  private final Context context;
+  private Activity activity;
+
+  QuickActions(Context context) {
+    this.context = context;
+  }
+
+  void setActivity(Activity activity) {
+    this.activity = activity;
+  }
+
+  public Activity getActivity() {
+    return this.activity;
+  }
+
+  @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.N_MR1)
+  boolean isVersionAllowed() {
+    // We already know that this functionality does not work for anything
+    // lower than API 25 so we chose not to return error. Instead we do nothing.
+    return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1;
+  }
+
+  @Override
+  public void setShortcutItems(
+      @NonNull List<ShortcutItemMessage> itemsList, @NonNull Result<Void> result) {
+    if (isVersionAllowed()) {
+      ShortcutManager shortcutManager =
+          (ShortcutManager) context.getSystemService(Context.SHORTCUT_SERVICE);
+      List<ShortcutInfo> shortcuts = ShortcutItemMessageToShortcutInfo(itemsList);
+      ThreadPoolExecutor executor =
+          new ThreadPoolExecutor(0, 1, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+
+      executor.execute(
+          () -> {
+            boolean dynamicShortcutsSet = false;
+            try {
+              shortcutManager.setDynamicShortcuts(shortcuts);
+              dynamicShortcutsSet = true;
+            } catch (Exception e) {
+              // Leave dynamicShortcutsSet as false
+            }
+
+            final boolean didSucceed = dynamicShortcutsSet;
+
+            new Thread(
+                    new Runnable() {
+                      @Override
+                      public void run() {
+                        // Run whatever background code you want here.
+                        if (didSucceed) {
+                          result.success(null);
+                        } else {
+                          result.error(
+                              new FlutterError(
+                                  "quick_action_setshortcutitems_failure",
+                                  "Exception thrown when setting dynamic shortcuts",
+                                  null));
+                        }
+                      }
+                    })
+                .start();
```

Comment:
This used to be run on the main thread. I'm not 100% sure this is the correct way to handle moving it to the background thread. Would appreciate some feedback.

Reply:
I believe that TODO meant that we could just remove this whole redispatch, because the engine added support for calling `result` on threads other than the platform thread.

I would suggest we leave the UI redispatch code as-is for this migration and change the threading in a different PR though, to minimize churn.